### PR TITLE
feat(cli): improve releasebinding generate with smart defaults and update-in-place

### PR DESCRIPTION
**Pull request by VajiraPrabuddhaka**
## Purpose                                                                                                                                                                             

  - Derive `--target-env` and `--use-pipeline` automatically from the project's
    `deploymentPipelineRef` and the pipeline's root environment, making them
    optional for project/component scopes (still required for `--all`)
  - Update existing bindings in-place at their original file path instead of
    creating duplicates
  - Wire the output directory resolver into single-component mode so `--component`
    generates files in the same location as `--all` and `--project`
  - Rename the default output directory from `bindings/` to `release-bindings/`
    for clarity

## Checklist
- [ ] Tests added or updated (unit, integration, etc.)
- [ ] Samples updated (if applicable)

## Changed Files Breakdown

**By top-level folder:**
- internal/: 7 files (+359/-69 lines) — internal/occ changes: core binding generation, index lookup, output writer, CLI releasebinding command, resolver and tests
  - Files: internal/occ/cmd/releasebinding/releasebinding.go, internal/occ/cmd/releasebinding/resolver.go, internal/occ/cmd/releasebinding/resolver_test.go, internal/occ/fsmode/index.go, internal/occ/fsmode/output/writer.go, internal/occ/fsmode/generator/release_binding.go, internal/occ/fsmode/generator/release_binding_test.go
- pkg/: 1 file (+7/-12 lines) — CLI flag handling (pkg/cli/cmd/releasebinding/generate.go)

Total: 8 files, approximately +366/-81 lines

---

## API / CRD Surface Changes

Compatibility risk: Low–Medium

- Added exported Index method:
  - Index.GetProject(namespace, name string) (*index.ResourceEntry, bool) — exposes project lookup (internal/occ/fsmode/index.go). Risk: Low (additive).
- Binding generation API surface:
  - BindingGenerator.GenerateBindingWithInfo(opts BindingOptions) (*BindingInfo, error) — returns BindingInfo with metadata including IsUpdate and ExistingFilePath (internal/occ/fsmode/generator/release_binding.go). Risk: Low (additive).
  - BindingInfo.ExistingFilePath string — new field to carry original file path for in-place updates. Risk: Low (additive).
- Writer/CLI behavior surface:
  - BulkBindingWriteOptions.ExistingPaths map[string]string — new field used by writer to prefer in-place writes (internal/occ/fsmode/output/writer.go). Risk: Medium (alters write semantics).
- Internal function changes (internal consumers only):
  - generateForComponent(...) now accepts resolver output.OutputDirResolverFunc and returns BindingInfo (internal/occ/cmd/releasebinding/releasebinding.go). Risk: Medium for internal callers; no public API break.

Summary: Additive public API surface changes are low risk; behavioral changes around write paths and CLI defaults are higher operational risk.

---

## Tests Added / Updated

Added:
- internal/occ/fsmode/generator/release_binding_test.go (+~196 lines)
  - TestGenerateBindingWithInfo_Create — asserts create path (IsUpdate=false, ExistingFilePath empty)
  - TestGenerateBindingWithInfo_Update — asserts update path (IsUpdate=true, ExistingFilePath set)
  - Helpers: newTestPipelineInfo, addRelease, addReleaseBinding

Updated:
- internal/occ/cmd/releasebinding/resolver_test.go — updated expectations to use release-bindings/ (small changes)

Coverage gaps / critical missing tests:
- No tests for CLI flag derivation logic that auto-derives --use-pipeline and --target-env from project/pipeline (releasebinding.go).
- No tests for dry-run output formatting that now prints bindingInfo.Binding YAML and actionCreated/actionUpdated.
- No end-to-end tests for write-in-place behavior (ExistingPaths precedence → actual file writes).
- No tests for full precedence chain resolution (ExistingPaths → --output-path → release-config → resolver → default).
- No integration/E2E tests validating create → update flow writing to original file paths and failure/rollback behavior.

---

## Behavioral Changes & Risk Hotspots

1) Flag auto-derivation (Medium)
- What: For project/component scopes, --use-pipeline and --target-env can be derived automatically from the project's deploymentPipelineRef and the pipeline's root environment; --all still requires --use-pipeline.
- Risk: Medium — implicit defaults may surprise users; validation moved into releasebinding logic and lacks targeted tests.

2) Update-in-place writes (High)
- What: Existing release bindings detected are updated in-place at their original file path (BindingInfo.ExistingFilePath plus ExistingPaths mapping used by writer).
- Risk: High — destructive modification of existing files without explicit rollback or confirmation; limited test coverage.

3) Output path resolution precedence (High)
- What: New precedence: ExistingPaths (explicit update mapping) → CLI OutputDir (--output-path) → release-config.getBindingOutputDir → Resolver function → fallback default (renamed to release-bindings/).
- Risk: High — complex precedence can cause unexpected write locations; interactions untested.

4) Default directory rename (Medium)
- What: Default directory renamed from bindings/ → release-bindings/ across resolver, writer, and tests.
- Risk: Medium — breaking for downstream scripts/CI expecting previous path; no migration note included.

5) Resolver wiring & signature change (Medium)
- What: OutputDirResolverFunc propagated into single-component generation so --component writes to same locations as --project/--all; generateForComponent signature updated accordingly.
- Risk: Medium — internal callsite changes performed, but behavior may have edge-case differences; limited tests.

6) Namespace requirement & pipeline validation (Low–Medium)
- What: GenerateReleaseBinding now requires namespace from context; derives/validates pipeline and target env and errors if missing.
- Risk: Low–Medium — stricter validation may surface new errors; no dedicated tests.

---

## Recommended Next Actions

- Add unit tests for:
  - releasebinding.go: flag derivation, namespace extraction, pipeline validation, dry-run output and action labeling.
  - writer.go: ExistingPaths precedence, conflict scenarios, resolver fallbacks, and error handling.
- Add integration/E2E test(s) exercising create → update flows to verify in-place file updates and compare dry-run vs actual write behavior.
- Add migration notes/documentation for default directory rename (bindings → release-bindings) for downstream users and CI.
- Consider adding an explicit confirmation flag, safety/rollback mechanism, or non-destructive default for in-place updates to reduce risk.

### CODE REVIEW

File: internal/occ/fsmode/output/writer.go [hunk 1]
```diff
@@ -495,18 +496,14 @@ func (w *Writer) resolveBindingOutputPath(
...
-			// If config path is relative, resolve it against baseDir
-			if !filepath.IsAbs(configDir) {
-				configDir = filepath.Join(w.baseDir, configDir)
-			}
-			return filepath.Join(configDir, bindingName+".yaml")
+	// Priority 0: If an existing file path is known (update-in-place), return it directly
+	if opts.ExistingPaths != nil {
+		if existingPath, ok := opts.ExistingPaths[bindingName]; ok && existingPath != "" {
+			return existingPath
 		}
```
[the remark this answers is not in the document]
Comment (VajiraPrabuddhaka) [reply internal/occ/fsmode/output/writer.go:503]: no this is fine, all the updating paths are relative to the root-dir (provided through arguments)

File: internal/occ/cmd/releasebinding/releasebinding.go [hunk 2]
```diff
@@ -228,33 +256,45 @@ func (r *ReleaseBindingImpl) generateForComponent(gen *generator.BindingGenerato
...
+		// CREATE: --output-path → release-config → resolver → default
+		if params.OutputPath != "" {
+			componentOutputDir = params.OutputPath
+		} else if releaseConfig != nil {
+			componentOutputDir = releaseConfig.GetBindingOutputDir(params.ProjectName, params.ComponentName)
+		}
+		if componentOutputDir == "" && resolver != nil {
+			componentOutputDir = resolver(params.ProjectName, params.ComponentName)
+		}
 	}
```
[the remark this answers is not in the document]
Comment (VajiraPrabuddhaka) [reply internal/occ/cmd/releasebinding/releasebinding.go:272]: This is ok, as occ treats all the paths here relative to the root-dir (which is a use input as an argument)

File: internal/occ/fsmode/generator/release_binding.go [hunk 3]
```diff
@@ -111,6 +112,36 @@ func (g *BindingGenerator) GenerateBinding(opts BindingOptions) (*unstructured.U
...
+
+	// Check if this is an update and capture existing file path
+	entry, exists := g.index.GetReleaseBindingForEnv(opts.ProjectName, opts.ComponentName, opts.TargetEnv)
+	if exists {
+		info.IsUpdate = true
+		info.ExistingFilePath = entry.FilePath
+	}
+
+	return info, nil
+}
```
[the remark this answers is not in the document]
Comment (VajiraPrabuddhaka) [reply internal/occ/fsmode/generator/release_binding.go:149]: added

File: internal/occ/cmd/releasebinding/releasebinding.go [hunk 4]
```diff
@@ -112,13 +116,37 @@ func (r *ReleaseBindingImpl) GenerateReleaseBinding(params api.GenerateReleaseBi
...
+		if pipelineRef == "" {
+			return fmt.Errorf("project %q has no deploymentPipelineRef set", params.ProjectName)
+		}
+		params.UsePipeline = pipelineRef
+		fmt.Printf("No --use-pipeline specified, using project's deploymentPipelineRef: %s\n", pipelineRef)
+	}
+
+	if params.UsePipeline == "" {
+		return fmt.Errorf("--use-pipeline is required (could not derive from project)")
+	}
```
[the remark this answers is not in the document]
Comment (VajiraPrabuddhaka) [reply internal/occ/cmd/releasebinding/releasebinding.go:135]: added

File: internal/occ/fsmode/generator/release_binding.go [hunk 5]
```diff
@@ -111,6 +112,36 @@ func (g *BindingGenerator) GenerateBinding(opts BindingOptions) (*unstructured.U
 	return g.buildMinimalBinding(opts.ProjectName, opts.ComponentName, releaseName, opts.TargetEnv, opts.Namespace), nil
 }
 
+// GenerateBindingWithInfo generates a single ReleaseBinding and returns rich info
+// including whether this is an update and the existing file path.
+func (g *BindingGenerator) GenerateBindingWithInfo(opts BindingOptions) (*BindingInfo, error) {
+	binding, err := g.GenerateBinding(opts)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (VajiraPrabuddhaka) [reply internal/occ/fsmode/generator/release_binding.go:127]: fixed

File: internal/occ/cmd/releasebinding/releasebinding.go [hunk 6]
```diff
@@ -129,15 +157,15 @@ func (r *ReleaseBindingImpl) GenerateReleaseBinding(params api.GenerateReleaseBi
...
 	}
 
-	// 6. Get namespace from context
-	namespace := ctx.Namespace
-	if namespace == "" {
-		return fmt.Errorf("namespace is required in context")
+	// Validate target environment exists in pipeline
+	if err := pipelineInfo.ValidateEnvironment(params.TargetEnv); err != nil {
+		return fmt.Errorf("invalid target environment: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (VajiraPrabuddhaka) [reply internal/occ/cmd/releasebinding/releasebinding.go:160]: fixed

File: internal/occ/cmd/releasebinding/derive_pipeline_test.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
+// Copyright 2025 The OpenChoreo Authors
```
Comment (JanakaSandaruwan) [inline internal/occ/cmd/releasebinding/derive_pipeline_test.go]:
```suggestion
// Copyright 2026 The OpenChoreo Authors
```

File: internal/occ/cmd/releasebinding/releasebinding.go [hunk 8]
```diff
@@ -112,13 +117,24 @@ func (r *ReleaseBindingImpl) GenerateReleaseBinding(params api.GenerateReleaseBi
 
 	ocIndex := fsmode.WrapIndex(persistentIndex.Index)
 
-	// 3. Load release config (optional - when absent, output dirs are inferred from index)
+	// 3. Get namespace from context
+	namespace := ctx.Namespace
+	if namespace == "" {
```
Comment (JanakaSandaruwan) [inline internal/occ/cmd/releasebinding/releasebinding.go:122]: The decided behaviour is to have a `--namespace` flag
1. if flag exists, it has priority
2. if not check in the context

File: internal/occ/fsmode/generator/release_binding_test.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
+// Copyright 2025 The OpenChoreo Authors
```
Comment (JanakaSandaruwan) [inline internal/occ/fsmode/generator/release_binding_test.go]:
```suggestion
// Copyright 2026 The OpenChoreo Authors
```

File: internal/occ/cmd/releasebinding/releasebinding.go (hunk 8 above)
Comment (VajiraPrabuddhaka) [reply internal/occ/cmd/releasebinding/releasebinding.go:122]: yes, will send this in a separate PR as an improvement

File: internal/occ/fsmode/generator/release_binding_test.go (hunk 9 above)
Comment (VajiraPrabuddhaka) [reply internal/occ/fsmode/generator/release_binding_test.go]: fixed

File: internal/occ/cmd/releasebinding/derive_pipeline_test.go (hunk 7 above)
Comment (VajiraPrabuddhaka) [reply internal/occ/cmd/releasebinding/derive_pipeline_test.go]: fixed

**Review by JanakaSandaruwan — APPROVED** [review]
